### PR TITLE
Fix bug in MassPay vis-a-vis refs

### PR DESCRIPTION
**Pull request by whit537**
Discovered in gratipay/inside.gratipay.com#1110.

### CODE REVIEW

**Comment by whit537** [comment]
Denied payments have an empty transaction id. Evident in the [log](https://github.com/gratipay/logs/blob/master/masspay/2017-06-29.report.paypal.csv) and appearing under inspection.

```diff
diff --git a/bin/masspay.py b/bin/masspay.py
index 99d7f8d..01715cd 100755
--- a/bin/masspay.py
+++ b/bin/masspay.py
@@ -179,8 +179,12 @@ def load_statuses_and_refs():
         if line.startswith('Transaction ID,Recipient'):
             break
     for rec in csv.reader(fp):
-        statuses[rec[1]] = _status_map[rec[5]]
-        refs[rec[1]] = rec[0]
+        ref = rec[0]
+        email = rec[1]
+        status = rec[5]
+        statuses[email] = _status_map[status]
+        refs[email] = ref
+    import pdb; pdb.set_trace()
     return statuses, refs
```

**Comment by whit537** [comment]
The MassPay itself has a unique transaction ID. We don't have access to that?

screen shot 2017-07-06 at 7 58 49 am

**Comment by whit537** [comment]
No, doesn't appear in the log.

We could manually add before posting masspay back, but we currently require a unique ref (if not `NULL`).

**Comment by whit537** [comment]
We could munge it to be unique. `1234567898765-a` or whatever.

**Comment by whit537** [comment]
Ready for review @rohitpaulk @clone1018.

**Comment by whit537** [comment]
This requires the payday runner to paste in the unique transaction id for the MassPay. I'll update the docs ...

File: bin/masspay.py [hunk 1]
```diff
@@ -176,16 +176,26 @@ def load_statuses_and_refs():
...
         if line.startswith('Transaction ID,Recipient'):
+            # This is the header row for the actual CSV data.
             break
-    for rec in csv.reader(fp):
-        statuses[rec[1]] = _status_map[rec[5]]
-        refs[rec[1]] = rec[0]
+    for i, rec in enumerate(csv.reader(fp)):
+        ref = rec[0]
+        if not ref:
+            ref = '{}-{:>04}'.format(masspay_id, i)
```
Comment (rohitpaulk) [inline bin/masspay.py]: Worth adding a comment that says why we're doing this?

Comment (whit537) [inline bin/masspay.py]: Done in dda94ad1de1ce74afe4108cd7e2278713b3a1b34.

**Comment by whit537** [comment]
Docs in https://github.com/gratipay/inside.gratipay.com/pull/1113.

**Comment by whit537** [comment]
> This requires the payday runner to paste in the unique transaction id for the MassPay.

Actually, this appears to not be true.

```
gratipay::DATABASE=> \d exchanges
                                     Table "public.exchanges"
┌─────────────┬──────────────────────────┬────────────────────────────────────────────────────────┐
│   Column    │           Type           │                       Modifiers                        │
├─────────────┼──────────────────────────┼────────────────────────────────────────────────────────┤
│ id          │ integer                  │ not null default nextval('exchanges_id_seq'::regclass) │
│ timestamp   │ timestamp with time zone │ not null default now()                                 │
│ amount      │ numeric(35,2)            │ not null                                               │
│ fee         │ numeric(35,2)            │ not null                                               │
│ participant │ text                     │ not null                                               │
│ recorder    │ text                     │                                                        │
│ note        │ text                     │                                                        │
│ status      │ exchange_status          │ not null                                               │
│ route       │ bigint                   │ not null                                               │
│ ref         │ text                     │                                                        │
└─────────────┴──────────────────────────┴────────────────────────────────────────────────────────┘
Indexes:
    "exchanges_pkey" PRIMARY KEY, btree (id)
Foreign-key constraints:
    "exchanges_participant_fkey" FOREIGN KEY (participant) REFERENCES participants(username) ON UPDATE CASCADE ON DELETE RESTRICT
    "exchanges_recorder_fkey" FOREIGN KEY (recorder) REFERENCES participants(username) ON UPDATE CASCADE ON DELETE RESTRICT
    "exchanges_route_fkey" FOREIGN KEY (route) REFERENCES exchange_routes(id)
```

**Comment by whit537** [comment]
I guess it's because it wants to be unique per network, and that's in a different table (`exchange_routes``)?

**Comment by whit537** [comment]
Gonna run this per @rohitpaulk in [slack](https://gratipay.slackarchive.io/gratipay/page-100/ts-1499354718691814):

> Looks good to me, aside from adding a small comment

Will leave the merge though ...

**Comment by kaguillera** [comment]
And it [ref] is needed because want to have a track it. This is part of backfill

**Comment by whit537** [comment]
Yep. Backfill being https://github.com/gratipay/gratipay.com/issues/4442.

**Pull request merged**